### ISSUE TITLE
Sampling methods

### ISSUE THREAD

**Issue by xyda** [clab/dynet#1067]
Does dynet provide random sampling functions such as sampling an element from a distribution of softmax?

**Comment by duyvuleo** [comment]
Softmax-Builders may be an option:
http://dynet.readthedocs.io/en/latest/builders.html#Softmax-Builders.

**Comment by xyda** [comment]
Thanks. The parallel implementation of sampling such as batched sampling would be great.

**Comment by redpony** [comment]
There is a long discussion about this. We realize we need to figure out a way of generating samples (or maxes, or ...) quickly without copying values from the GPU to the CPU or between different parts of CPU memory. Unfortunately, due to DyNet's design simplicity (everything inside DyNet is a differentiable function), things that return non-differentiable values (e.g., discrete indices or symbols) don't really have a good place to live. We  need a general way of adding non-differentiable forward-triggering operations (e.g., forward(), scalar_value(), npvalue(), ...) so requests like this can be handled, but we haven't yet designed this.

**Comment by xyda** [comment]
Thanks for the explanation. I noticed that in tensor.h, there is an implementation of parallel sampling function:

static IndexTensor categorical_sample_log_prob(const Tensor& v, unsigned dim = 0, unsigned num = 1);

Is it the correct function that I can use? Although it is non-differentiable.

**Comment by neubig** [comment]
Yes, this will work, but it's somewhat experimental. We're currently designing a more permanent interface and will update here when we do.
